### DISCUSSION TITLE
Support multiple --netmask filters?

### DISCUSSION THREAD

**Discussion by drws** [aircrack-ng/aircrack-ng#2464] [Ideas]
The ability to match multiple `--bssid` values has been recently added to `airodump-ng` (issue #1940, pull #2432).

Also a `--netmask` argument can be applied over BSSID(s) for _wildcard_ matching.

The idea to discuss here is how `airodump-ng` should handle not only single, but multiple netmasks as well. Furthermore, how an arbitrary number of BSSID(s)-mask(s) pairs could be handled - input parsing and functionally?

A minimal example: matching a single BSSID and another masked one (in the same run).

**Comment by drws** [comment]
My proposal based on the minimal example above:

`airodump-ng --bssid $B1 --bssid2 $B2 --netmask2 $M2`

`--bssid` would only be an alias for `--bssid1` and similarly `--netmask`. `--bssidN` and `--netmaskN` need a bit of input parsing trickery, but they should allow for an arbitrary input while keeping everything sane.

Also multiple items could be added to the Nth list by passing the same argument more than once. For example:

`airodump-ng --bssid1 $B1 --bssid1 $B2 --netmask1 $M1 --netmask1 $M2 ...`

Since the values are in the same list no. 1, all combinations are to be tested:

    $B1 ^ $M1
    $B1 ^ $M2
    $B2 ^ $M1
    $B2 ^ $M2

**Reply by Mister-X-** [reply]
I'm not really sure we can have dynamic parameters without a lot of extra code. So, we'll have to stick with `--bssid` and `--netmask`.

I got to some thinking, and I think we could do by position, where the `--netmask` refers to the last `--bssid`, as `getopt()` [returns parameters in order](https://stackoverflow.com/questions/29453123/how-to-determine-order-and-count-of-arguments-in-getopt). It's *rather* easy in the code as we match the current netmask with the last **current** BSSID, and also easy to understand (and makes sense) from a user's perspective.

So, for example, with:

`airodump-ng --bssid $BSSID1 --bssid $BSSID2 --netmask $MASK2 --bssid `

The first `$BSSID` doesn't have a netmask, and `$BSSID2` has `$MASK1`.

**Reply by drws** [reply]
Dynamic parameters could technically be worked around by passing the list ID in the argument value (`--bssid 1+$B1 ...`), but that doesn't help code complexity. Implementing lists is a less simple solution, but it provides more flexibility.

While your solution is compact its input syntax isn't as much. It suits well the minimal example, but what if user wants to apply a single mask to numerous BSSIDs? Lots of unnecessary mask repetitions are needed, for example.

**Comment by ZerBea** [comment]
How about using a Berkeley Packet Filter on the packet socket instead of walking through filter lists.
`setsockopt(fd_socket_rx, SOL_SOCKET, SO_ATTACH_FILTER, &bpf, sizeof(bpf))
`
